### PR TITLE
Fix for Sinon.JS issue #766: clearTimeout() no longer clears Immediat…

### PR DESCRIPTION
**Pull request by rogierschouten**
…e/Interval and vice versa

A clearTimeout(), given the ID of a timer started with setImmediate() or
setInterval() would erroneously clear that immediate or interval

### CODE REVIEW

**Comment by mantoni** [comment]
This can't be merged for some reason.

**Comment by rogierschouten** [comment]
Can you describe what happens? It says "This branch is up-to-date with the base branch" here and I don't know what to do

**Comment by mantoni** [comment]
Funny. It's resolved. Can be merged again.

File: src/lolex.js [hunk 1]
```diff
@@ -258,6 +258,42 @@
...
+
+    function clearTimer(clock, timerId, ttype) {
+        if (!timerId) {
+            // null appears to be allowed in most browsers, and appears to be
+            // relied upon by some libraries, like Bootstrap carousel
+            return;
+        }
+
+        if (!clock.timers) {
+            clock.timers = [];
```
Comment (mantoni) [inline src/lolex.js]: You could simply return here.

Comment (rogierschouten) [inline src/lolex.js]: I agree, but I did not change that code. I moved that entire function body from its old place in clock.clearTimeout() and only added the last few lines.  
That thing with setting clock.timers to [] was so specific that I thought it must have a function and I didn't want to mess with it. Want me to change it anyway?

Comment (mantoni) [inline src/lolex.js]: Ah, I see. Leave it as it is then.

**Comment by mantoni** [comment]
Ok, now this is odd: I ran a couple of tests and this is how it behaves in the real world.

### Node 0.10.38

```js
var t = setTimeout(function () { console.log('!'); }, 10); clearInterval(t); // Prints !
var t = setTimeout(function () { console.log('!'); }, 10); clearImmediate(t); // Doesn't print !

var t = setImmediate(function () { console.log('!'); }); clearTimeout(t); // Prints !
var t = setImmediate(function () { console.log('!'); }); clearInterval(t); // Also prints !
```

### Chrome 43

```js
var t = setTimeout(function () { console.log('!'); }, 10); clearInterval(t); // Doesn't print !
// setImmediate and clearImmediate throw because they're not available.
```

### Firefox 31.2.0

```js
var t = setTimeout(function () { console.log('!'); }, 10); clearInterval(t); // Doesn't print !
// setImmediate and clearImmediate throw because they're not available.
```

### IE 9

```js
var t = setTimeout(function () { console.log('!'); }, 10); clearInterval(t); // Doesn't print !
// setImmediate and clearImmediate throw because they're not available.
```

That's all I have at hand at the moment. To my surprise, it turns out the current Lolex behavior is closer to reality than it would be with your fix. However, node 0.10 is even inconsistent in it's behavior. I would be very interested in how io.js behaves. I can check that later at home.

**Comment by rogierschouten** [comment]
That is true, but in this case you don't want to mimick reality (I had this discussion on the sinon issue as well). It is bad programming practice to have setTimeout() followed by clearImmediate() -  and on some platforms, as you saw, it will not do what you intended. Therefore it would be handy if Sinon was strict so you can find your bug in unittests instead of discovering it on one of the platforms

**Comment by mantoni** [comment]
Reasonable. Shouldn't it then even throw if you're trying to, let's say, clear an interval with `clearTimeout`? With this change it would just silently do nothing.

**Comment by rogierschouten** [comment]
Good idea, done!

**Comment by cjohansen** [comment]
Lolex 1.3.0 out

**Comment by cjohansen** [comment]
Sinon 1.16.0 out

**Comment by rogierschouten** [comment]
Awesome, thanks!!

**Comment by mantoni** [comment]
:+1:

**Pull request merged**